### ISSUE TITLE
Close connection after optimize

### ISSUE THREAD

**Issue by datasatanic** [optuna/optuna#1347]
## Expected behavior

I need to build many, fast-training models. For this i'm using multiprocessing.Pool:
```
with Pool(maxtasksperchild=10) as pool:
    res=pool.map(task,iter)
```
In each task i'm creating stydy with unique name  and run optimize. As storage is used Postgres 12.

After a few cycle, i get error from psycopg2: "too many client"
How I can manualy close connection for finished task and study. Or exist anotherway?
P.S. I know about the possibility to шncrease quantity of clients on Postgres, but It only test data. Real task is much more.

## Environment

- Optuna version: 1.5.0
- Python version: 3.7.7
- OS: Ubuntu 20.04
-  psycopg2-binary: 2.8.5
- Postgres version: 12

## Error messages, stack traces, or logs

```
# error messages, stack traces, or logs
The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/ilya/anaconda3/lib/python3.7/multiprocessing/pool.py", line 121, in worker
    result = (True, func(*args, **kwds))
  File "/home/ilya/anaconda3/lib/python3.7/multiprocessing/pool.py", line 44, in mapstar
    return list(map(*args))
  File "<ipython-input-11-68b9f5f269e0>", line 3, in task
    df:pd.DataFrame=pd.read_sql_query(select_query.format(pan=pan),connect_string)
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/pandas/io/sql.py", line 332, in read_sql_query
    chunksize=chunksize,
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/pandas/io/sql.py", line 1218, in read_query
    result = self.execute(*args)
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/pandas/io/sql.py", line 1087, in execute
    return self.connectable.execute(*args, **kwargs)
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 2243, in execute
    connection = self._contextual_connect(close_with_result=True)
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 2311, in _contextual_connect
    self._wrap_pool_connect(self.pool.connect, None),
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 2349, in _wrap_pool_connect
    e, dialect, self
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 1591, in _handle_dbapi_exception_noconnection
    sqlalchemy_exception, with_traceback=exc_info[2], from_=e
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 178, in raise_
    raise exception
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/engine/base.py", line 2345, in _wrap_pool_connect
    return fn()
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/pool/base.py", line 364, in connect
    return _ConnectionFairy._checkout(self)
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/pool/base.py", line 778, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/pool/base.py", line 495, in checkout
    rec = pool._do_get()
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/pool/impl.py", line 140, in _do_get
    self._dec_overflow()
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/util/langhelpers.py", line 69, in __exit__
    exc_value, with_traceback=exc_tb,
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 178, in raise_
    raise exception
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/pool/impl.py", line 137, in _do_get
    return self._create_connection()
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/pool/base.py", line 309, in _create_connection
    return _ConnectionRecord(self)
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/pool/base.py", line 440, in __init__
    self.__connect(first_connect_check=True)
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/pool/base.py", line 661, in __connect
    pool.logger.debug("Error on connect(): %s", e)
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/util/langhelpers.py", line 69, in __exit__
    exc_value, with_traceback=exc_tb,
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/util/compat.py", line 178, in raise_
    raise exception
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/pool/base.py", line 656, in __connect
    connection = pool._invoke_creator(self)
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/engine/strategies.py", line 114, in connect
    return dialect.connect(*cargs, **cparams)
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/sqlalchemy/engine/default.py", line 490, in connect
    return self.dbapi.connect(*cargs, **cparams)
  File "/home/ilya/anaconda3/lib/python3.7/site-packages/psycopg2/__init__.py", line 127, in connect
    conn = _connect(dsn, connection_factory=connection_factory, **kwasync)
sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) FATAL:  sorry, too many clients already

(Background on this error at: http://sqlalche.me/e/e3q8)
```

**Comment by datasatanic** [comment]
``` study._storage._backend.engine.dispose()``` not help

**Comment by ytsmiling** [comment]
Thank you for reporting the problem. I remember that I encountered a similar problem, but I somehow failed to reproduce the problem. Would you mind sharing code with the problem?

**Comment by datasatanic** [comment]
Example of code:
```python
import pandas as pd
import numpy as np
import optuna

from multiprocessing import Pool


def obj_func(trial:optuna.Trial,arg1:float):
    x=trial.suggest_float('x',0,100)
    return arg1*(x-3.3)**2



obj_func(optuna.trial.FixedTrial({'x':3.2}),1)


def task(arg1):
    study:optuna.Study=optuna.create_study(conn_str,study_name=f'test_{arg1}')
    study.optimize(lambda x:obj_func(x,arg1),n_trials=10)
    return study


optuna.logging.disable_default_handler()
optuna.logging.set_verbosity(optuna.logging.ERROR)


with Pool(maxtasksperchild=10) as pool:
    res=pool.map(task,np.random.rand(1000))









```

**Comment by toshihikoyanase** [comment]
Thank you for sharing your code. I reproduced the problem with the master branch (https://github.com/optuna/optuna/commit/05de3f7ff673364ca822fd2119a80dd37a04b9cb).

I'm sorry for the inconvenience, but we may not have enough resources to investigate this problem in detail because members of the core development team focus on the [v2 release tasks](https://github.com/optuna/optuna/issues/1246).
We'll revisit this problem after the release.

**Comment by hvy** [comment]
I believe a solution would be require some discussions but there seem to be two issues.

**Issue 1**

Each `Study` instance holds its own `RDBStorage`, which holds its own `sqlalchemy` engine (and implicitly its own connection pool). When a study is instantiated, at least one connection is established and then placed back into the pool. This means that creating more studies than allowed connections (default 100 I believe in the case of postgresql) will lead the same error described by the author.

```python
import optuna

conn_str= "postgresql://localhost:5432/foo"

studies = []
for i in range(101):  # 100 is the maximum allowed number of connections for postgresql by default.
    study = optuna.create_study(conn_str, study_name=f"test_{i}")  # Error when `i == 100`.
    study._storage._backend.delete_study(study._study_id)
    studies.append(study)
 ```

In the initial example, since we're returning (pickling/unpickling) studies from each worker, running more jobs than the number of maximum allowed connections will lead to this error. You must call `dispose` [here when the study is unpickled/returned](https://github.com/optuna/optuna/blob/master/optuna/study.py#L186).

**Issue 2**

The connections pools aren't disposed when the studies are deleted. So after each call to `task` returns, a connections is left established to the database. This can be worked around by calling `dispose` before returning from `task`. I actually tried to address this by calling `dispose` in the [`_finalize` method](https://github.com/optuna/optuna/blob/master/optuna/storages/_rdb/storage.py#L119) of the `RDBStorage` but noticed that this finalizer registration might be buggy. It holds a reference to the object itself so it might not get deleted at all. Read the [note at the bottom](https://docs.python.org/3/library/weakref.html#weakref.finalize).

**TLDR/Possible fix**

A workaround is to call `dispose` on the storage belonging to a study that's no longer needed, and additionally when a new e.g. unpickled study is created. I believe a proper fix would be to share engines between multiple storage objects but this might be a breaking change.

**Comment by hvy** [comment]
Some discussion have been going on how and when to tackle this issue. I don't think it's a hurry but, @keisuke-umezawa were you interested?

**Comment by keisuke-umezawa** [comment]
@hvy Thank you for addressing to me! I will work on it. At first, I will create multiple plans and designs for fixing the bug.

**Comment by keisuke-umezawa** [comment]
## Issues

> **Issue 1**
> 
> Each `Study` instance holds its own `RDBStorage`, which holds its own `sqlalchemy` engine (and implicitly its own connection pool). When a study is instantiated, at least one connection is established and then placed back into the pool. This means that creating more studies than allowed connections (default 100 I believe in the case of postgresql) will lead the same error described by the author.
> 
> ```python
> import optuna
> 
> conn_str= "postgresql://localhost:5432/foo"
> 
> studies = []
> for i in range(101):  # 100 is the maximum allowed number of connections for postgresql by default.
>     study = optuna.create_study(conn_str, study_name=f"test_{i}")  # Error when `i == 100`.
>     study._storage._backend.delete_study(study._study_id)
>     studies.append(study)
>  ```
> 
> In the initial example, since we're returning (pickling/unpickling) studies from each worker, running more jobs than the number of maximum allowed connections will lead to this error. You must call `dispose` [here when the study is unpickled/returned](https://github.com/optuna/optuna/blob/master/optuna/study.py#L186).
> 
> **Issue 2**
> 
> The connection pools aren't disposed when the studies are deleted. So after each call to `task` returns, a connections is left established to the database. This can be worked around by calling `dispose` before returning from `task`. I actually tried to address this by calling `dispose` in the [`_finalize` method](https://github.com/optuna/optuna/blob/master/optuna/storages/_rdb/storage.py#L119) of the `RDBStorage` but noticed that this finalizer registration might be buggy. It holds a reference to the object itself so it might not get deleted at all. Read the [note at the bottom](https://docs.python.org/3/library/weakref.html#weakref.finalize).

Actually, to run this code https://github.com/optuna/optuna/issues/1347#issuecomment-645201744, we have another issue here, even if we `share engines between multiple storage objects` in one process.

**Issue 3**
When we run tasks on different processes, the pooled connections are not shared with different processes. c.f. https://docs.sqlalchemy.org/en/13/core/pooling.html#using-connection-pools-with-multiprocessing-or-os-fork

## Actual use cases of processes and studies
Thinking about how optuna is used, the followings are candidates of use cases:

1. One process, with one running study instance
2. ~One process, with multiple running study instances~
3. Multiple process, with one study instance for each process

Regarding to actual use cases, we rarely run optuna on `One process, with multiple running study instances`, so we only need to take care 1 and 3 in the above.

## Solutions
### for 1 `One process, with one running study instance`
We need to solve the issue 2. We need to dispose unused connections.

### for 3 `Multiple process, with one study instance for each process`
We need to solve the issue 3. We need to share connection pools with multiprocess.

**Comment by keisuke-umezawa** [comment]
If you run the code in a single process, you can share a storage and a engine by the following code:

```
import optuna

conn_str = "postgresql://localhost:5432/foo"

studies = []
storage = optuna.storages.get_storage(conn_str)
for i in range(101):  # 100 is the maximum allowed number of connections for postgresql by default.
    study = optuna.create_study(study_name=f"test_{i}", storage=storage)
    study._storage._backend.delete_study(study._study_id)
    studies.append(study)
```

**Comment by keisuke-umezawa** [comment]
@datasatanic 
Also, in multiprocess case, the following code can be run.

```
import optuna
from multiprocessing import Pool
import numpy as np


conn_str = "postgresql://localhost:5432/foo"
storage = optuna.storages.get_storage(conn_str)


def task(arg1):
    storage._backend.engine.dispose()
    study = optuna.create_study(storage=storage, study_name=f"test_{arg1}")
    study._storage._backend.delete_study(study._study_id)
    return study


if __name__ == "__main__":

    optuna.logging.disable_default_handler()
    optuna.logging.set_verbosity(optuna.logging.ERROR)

    with Pool(processes=10) as pool:
        res = pool.map(task, [i for i in range(101)])
```

Note: To share the engine with multiple processes, I follow the official guide as follows:

> Call Engine.dispose() on any given Engine as soon one is within the new process.

c.f. https://docs.sqlalchemy.org/en/13/core/pooling.html#using-connection-pools-with-multiprocessing-or-os-fork

**Comment by keisuke-umezawa** [comment]
Still, I will continue to work on it

> Issue 2
> 
> The connection pools aren't disposed when the studies are deleted. So after each call to task returns, a connections is left established to the database. This can be worked around by calling dispose before returning from task. I actually tried to address this by calling dispose in the _finalize method of the RDBStorage but noticed that this finalizer registration might be buggy. It holds a reference to the object itself so it might not get deleted at all. Read the note at the bottom.

**Comment by keisuke-umezawa** [comment]
I will close this issue, because the issue 1 and 2 were resolved.

The issue 1 was resolved by this workaround. https://github.com/optuna/optuna/issues/1347#issuecomment-732042329

And, the issue 2 was also resolved by #2079 .
